src/MacroExpander.js

```diff
@@ -25,18 +28,87 @@ MacroExpander.prototype.nextToken = function() {
+        var tok;
         var expansion = this.macros[name];
         if (typeof expansion === "string") {
+            var numArgs = 0;
+            if (expansion.indexOf("#") !== -1) {
+                var stripped = expansion.replace(/##/g, "");
+                while (stripped.indexOf("#" + (numArgs + 1)) !== -1) {
+                    ++numArgs;
+                }
+            }
             var bodyLexer = new Lexer(expansion);
             expansion = [];
-            var tok = bodyLexer.lex();
+            tok = bodyLexer.lex();
             while (tok.text !== "EOF") {
                 expansion.push(tok);
                 tok = bodyLexer.lex();
             }
             expansion.reverse(); // to fit in with stack using push and pop
+            expansion.numArgs = numArgs;
             this.macros[name] = expansion;
         }
+        if (expansion.numArgs) {
+            var args = [];
+            var i;
+            // obtain arguments, either single token or balanced {…} group
+            for (i = 0; i < expansion.numArgs; ++i) {
+                var startOfArg = this.get(true);
+                if (startOfArg.text === "{") {
+                    var arg = [];
+                    var depth = 1;
+                    while (depth !== 0) {
+                        tok = this.get(false);
+                        arg.push(tok);
+                        if (tok.text === "{") {
+                            ++depth;
+                        } else if (tok.text === "}") {
+                            --depth;
+                        } else if (tok.text === "EOF") {
+                            throw new ParseError(
+                                "End of input in macro argument",
+                                startOfArg);
+                        }
+                    }
+                    arg.pop(); // remove last }
+                    arg.reverse(); // like above, to fit in with stack order
+                    args[i] = arg;
+                } else if (startOfArg.text === "EOF") {
+                    throw new ParseError(
+                        "End of input expecting macro argument", topToken);
+                } else {
+                    args[i] = [startOfArg];
+                }
+            }
+            // paste arguments in place of the placeholders
+            expansion = expansion.slice(); // make a shallow copy
+            for (i = expansion.length - 1; i >= 0; --i) {
+                tok = expansion[i];
+                if (tok.text === "#") {
+                    if (i === 0) {
+                        throw new ParseError(
+                            "Incomplete placeholder at end of macro body",
+                            tok);
+                    }
+                    tok = expansion[--i]; // next token on stack
+                    if (tok.text === "#") { // ## → #
+                        expansion.splice(i + 1, 1); // drop first #
+                    } else if (/^[1-9]$/.test(tok.text)) {
```

Comment:
It's worse than that benchmark suggests (which is already 10x on a size-4 array): `shift`/`unshift`'s cost grow linearly with array size, so things will get worse and worse as if your stack grows at all, while `push`/`pop` take constant time (amortized).  So I'd avoid that.  Indexing instead of removal is a reasonable way to go, though.  (In general, it sounds like you're looking for a **deque** data structure.)